### ISSUE TITLE
Better parameter openapi resolution

### ISSUE THREAD

**Issue by JakubMifek** [strongloop/loopback-next#3730]
## Suggestion

I have a scenario where my microservice requires a header parameter that is used in a sequence but never in controllers. Now, when I want to document the API, I'd like the parameter to be mentioned in the explorer.

My idea was that I'd append the parameter specification to each of controllers' methods. But that fails because parameter specifications are overwritten when any `@param` method parameter is used.

I would like to change the openapi generation in `@loopback/openapi-v3/src/controller-spec.ts` so that it overwrittes specified parameters only when it has to.

## Examples

```ts
  @get('/test/{test2}', {
    parameters: [
      {
        name: 'test1',
        in: 'header',
        required: true,
        schema: { type: 'string' },
        description: 'Test 1',
      },
    ],
    responses: { },
  })
  async test(
    @param.path.string('test2') test2: string,
  ): Promise<void> {
    console.log(test2);
  }
```
I would like the generated openapi to include both `test1` and `test2` parameters. Not just `test2`. And I do not want to include `test1` parameter as the method `@param` argument since it is not used in the method itself.

## Proposal

Change:
```ts
operationSpec.parameters = params
        .filter(p => p != null)
        .map(p => {
          // Per OpenAPI spec, `required` must be `true` for path parameters
          if (p.in === 'path') {
            p.required = true;
          }
          return p;
        });
```

to 

```ts
operationSpec.parameters = operationSpec.parameters || [];
const ps = params
        .filter(p => p != null)
        .map(p => {
          // Per OpenAPI spec, `required` must be `true` for path parameters
          if (p.in === 'path') {
            p.required = true;
          }
          return p;
        });
const psNames = ps.map(p => p.name);
// The references should be better managed.
operationSpec.parameters = operationSpec.parameters.filter(p => p.hasOwnProperty('$ref') || !psNames.includes((p as ParameterObject).name)).concat(ps);
```

## Acceptance criteria

TBD - will be filled by the team.

**Comment by achrinza** [comment]
Does [Extending OpenAPI specification](https://loopback.io/doc/en/lb4/Extending-OpenAPI-specification.html) resolve the issue?

**Comment by JakubMifek** [comment]
Hi @achrinza thank you for taking interest.

I guess it would, assuming that the `spec` parameter is complete openapi specification containing all endpoints where I'd be able to find my concrete endpoints and extend their specification using the parameter.

In the meanwhile I already solved the issue (sorry for not posting) by creating a bootscript which runs after the application boot but before application starts listening for incoming connections. This bootscript takes all controllers from the application context, goes through their methods and extends their openapi parameter metadata where i need it :)

I will close the issue now since I solved it in my case and there seems to be a solution in loopback documentation using the `SpecEnhancer` class.
